### ISSUE TITLE
Escape character when exporting in SQL

### ISSUE THREAD

**Issue by antoine2711** [OpenRefine/OpenRefine#4716]
It would be good to be able to have data escaped with the SQL importer, and to be able to choose by which character.
This was discussed in Issue #1940

### Proposed solution
Add a text field for an escape character in the SQL Exporter.

### Alternatives considered
data manipulation at the text level after exportation.

### Additional context
![image](https://user-images.githubusercontent.com/12217525/162367781-4533c000-3686-4377-b2de-ad8374df0a09.png)

**Comment by the-star-sea** [comment]
Hello.I am a junior undergraduate cs student and I wanna have a try.Thank u

**Comment by elroykanye** [comment]
Hey @the-star-sea , please go ahead. I'll assign you to this! ☺️

**Comment by the-star-sea** [comment]
I have read the Issue #1940 .Could u please give me an example?I have read the Issue #1940 .So Does the text assign the unicode outside the text?For example,if the escape character is " and there is a column named abc,it would be "abc".Do I understand correctly?

**Comment by the-star-sea** [comment]
I have read #1940 .Could u please give me an example?So Does the text assign the unicode outside the text?For example,if the escape character is " and there is a column named abc,it would be "abc".Do I understand it correctly?

**Comment by antoine2711** [comment]
@the-star-sea: if you try to SQL export a table, this is the current result.

```
CREATE TABLE TEST_food_small (
NDB_No VARCHAR(255) NULL,
Shrt_Desc VARCHAR(255) NULL,
(…)
Refuse_Pct VARCHAR(255) NULL
);
```

With a text field in which the tick character would be entered i.e. « ` », it would give this output:

```
CREATE TABLE `TEST_food_small` (
`NDB_No` VARCHAR(255) NULL,
`Shrt_Desc` VARCHAR(255) NULL,
(…)
`Refuse_Pct` VARCHAR(255) NULL
);
```

Regards, Antoine

**Comment by sherrif10** [comment]
Hello @antoine2711  , this issue looks amazing for me to dive into it  kindly assigne it to me , probably highlighting key points would be highly appreciated.

**Comment by antoine2711** [comment]
> Hello @antoine2711 , this issue looks amazing for me to dive into it kindly assigne it to me

@sherrif10: unfortunately, someone is already « assigned » to the issue, @the-star-sea. Since it's only been a few days, he most probably still working the issue. Please choose another one, not assigned.

Regards, Antoine

**Comment by sherrif10** [comment]
Thank you so much for your great response , i appreciate, My plan is to push more code this week as much as possible.

**Comment by the-star-sea** [comment]
> > Hello @antoine2711 , this issue looks amazing for me to dive into it kindly assigne it to me
> 
> @sherrif10: unfortunately, someone is already « assigned » to the issue, @the-star-sea. Since it's only been a few days, he most probably still working the issue. Please choose another one, not assigned.
> 
> Regards, Antoine

Thank u antoine,I am still working on it.

**Comment by the-star-sea** [comment]
I have complete it!What should I do I do next?I have forked one repository many days ago.@[antoine2711](https://github.com/antoine2711)

**Comment by the-star-sea** [comment]
I have complete it.However,I use fork and change on it.I am a new bee of opensource project.What should I do next ?Thank u so much.

**Comment by elroykanye** [comment]
> I have complete it.However,I use fork and change on it.I am a new bee of opensource project.What should I do next ?Thank u so much.

Hi @the-star-sea , you should find some helpful tips from these links:

- [Contributing Guide](https://github.com/OpenRefine/OpenRefine/blob/master/CONTRIBUTING.md)
- [Your first pull request](https://docs.openrefine.org/technical-reference/contributing#your-first-code-pull-request)

**Comment by antoine2711** [comment]
> I have complete it.However,I use fork and change on it.I am a new bee of opensource project.What should I do next ?Thank u so much.

@the-star-sea: Go on [Gitter](https://gitter.im/OpenRefine/OpenRefine/), and ask some help with creating your full pull request. But only do so after having read all the technical documentation. This aspect is in part covered, be help sometimes helps after reading it.

Regards, Antoine

**Comment by wetneb** [comment]
Hi @antoine2711, is this really needed now that we are normalizing column names in the SQL export? Your proposal seem to be a different solution for a problem we have already solved another way, no?

**Comment by antoine2711** [comment]
> Hi @antoine2711, is this really needed now that we are normalizing column names in the SQL export? Your proposal seem to be a different solution for a problem we have already solved another way, no?

I personally have work scenario where the fields name contains spaces, and they must be exactly like that. With a SQL query per table, I can rename them all, but having the SQL ready makes it's more useful. It's a very simple change for a worth feature, in my opinion. I will use it. I think replacing and escaping are two good solutions, that why I keep working this feature.

Regards, Antoine

**Comment by wetneb** [comment]
Yes but then you need to make sure you can actually disable the replacing, which you have not mentioned in the issue so far, right?

**Comment by niyolynate** [comment]
Having read a good first issue, this is my contribution to this project. now so good as an Outreachy project applicant since I have not seen others participate. it though the interest I developed. still looking around for others. any review and guidance are welcome. 
![SQL Importer](https://user-images.githubusercontent.com/98440370/164453075-05b61579-f19e-4a2f-8eea-cbb975e53bd6.png)

**Comment by antoine2711** [comment]
> Yes but then you need to make sure you can actually disable the replacing, which you have not mentioned in the issue so far, right?

Yes, you are rigth to ploint out that both of these features make little sens togeter. It's more one or the other. But, having them both at the same time can't create problems. If the current contributor would have a good first draft, that would have been a suggestion. Maybe he would have done it by himself. For sure, coding this after is not more work in itself, so I would have guide the contributor through this process. I'm learning this, so if you have particular recommandation, please let me know. One of the reason to propose this feature was also because we were lacking first good issues. This seams in overall relatively simple to code. I'm all hear @wetneb, recommendations?

Regards, Antoine

**Comment by niyolynate** [comment]
ok Point taken

**Comment by niyolynate** [comment]
my question is, do i use the code base for this or or use the open, if
code, i have tried several times to clone the git openrefine but after edit
rejected pushing. can i know which way please..

**Comment by antoine2711** [comment]
Hi @niyolynate, your question was not very clear, but these one are, so I'll answers them.

Once you have clone the git on your mechine, yes, you should create a NEW branch. This, in turn, will generate your own PR for this task. But I would not advise you to do so, because someone is already assigned to this issue. 

![image](https://user-images.githubusercontent.com/12217525/164460991-8d47392e-b5bd-495d-8d71-ca2e01d63605.png)

You should find one that not taken. For starting, find one with the label `Good First Issue`, and preferably not the label `To Be Reviewed`. Hope that helps!

Regards, Antoine

**Comment by niyolynate** [comment]
yes, it is,   Glad I Asked. Thanks a lot.

**Comment by antoine2711** [comment]
> Yes but then you need to make sure you can actually disable the replacing, which you have not mentioned in the issue so far, right?

https://github.com/OpenRefine/OpenRefine/blob/bef50d989b3c4b4216a5b71e99d4d4187046d0c3/main/src/com/google/refine/exporters/sql/SqlCreateBuilder.java#L85-L94

@wetneb: I feel that in my enthousiam, I might have overlooked the Java part, which is, as you know, not my strengh. So, to better answers your question, yes, a new option would be necessary. What is involve more than sending it thru the IP with the others options? For the extra option, how about `escapeColNames`?

Also, there should be a discussion of not sending nothing if the escaping character is null. In that case, it would be very safe to use the caret `.

Yes, I have a bit of review work to do here, since I volunteered. ;-)

Regards, Antoine

**Comment by the-star-sea** [comment]
Actually I have a version without replacement merging into because this is
my first issue and I forget to fork before I complete it.I rewrite the code
after I fork and I have to change the code to make it harmony with the
replacement.


Antonin Delpeuch ***@***.***> 于 2022年4月21日周四 19:41写道：

**Comment by niyolynate** [comment]
in order to 

public void EnableButtons()
{
    btnImportHFR.Enabled = Convert.ToBoolean(Shared.FilesForImport.Count);
    btnImportHFR.Text = btnImportHFR.Enabled ? "Find Available HFR" : "No HFR For Import";

    btnGetFacilityStatus.Enabled = Shared.sqlWrap.GetDataForFacStat(Shared.DsFacStat);
    updateButtons = false;
    
         CREATE TABLE userid(
    id Country(30) NOT NULL PRIMARY KEY,
    pwd Country30) NOT NULL,
    fullname EnterName50),
    email EnterEmail(50) );

**Comment by niyolynate** [comment]
> > Yes but then you need to make sure you can actually disable the replacing, which you have not mentioned in the issue so far, right?
> 
> https://github.com/OpenRefine/OpenRefine/blob/bef50d989b3c4b4216a5b71e99d4d4187046d0c3/main/src/com/google/refine/exporters/sql/SqlCreateBuilder.java#L85-L94
> 
> @wetneb: I feel that in my enthusiasm, I might have overlooked the Java part, which is, as you know, not my strength. So, to better answers your question, yes, a new option would be necessary. What is involved more than sending it thru the IP with the others options? For the extra option, how about `escapeColNames`?
> 
> Also, there should be a discussion of not sending nothing if the escaping character is `null`. In that case, it would be very safe to use the caret `.
> 
> Yes, I have a bit of review work to do here, since I volunteered. ;-)
> 
> Regards, Antoine

yes, Your right. Still Learning.

**Comment by wetneb** [comment]
After a quick search, I found [this StackOverflow answer](https://stackoverflow.com/a/2901499/985087) which seems to suggest that the standard mandates double quotes to quote fields. THerefore, I do not think it is worth letting the user choose which quote character should be used.

A user could want:
* the current behaviour where column names are normalized
* or the preservation of the original column names, enclosed by double quotes

So I would say we could give the choice between those two options in the UI.

**Comment by antoine2711** [comment]
@the-star-sea: in regards to @wetneb last comment, I suggest this user interface. 
![image](https://user-images.githubusercontent.com/12217525/171880614-d2226fe3-cb3a-4c02-8a1f-f04c4b295682.png)

What do you think?

Regards, Antoine

**Comment by wetneb** [comment]
I am struggling to understand why we need three options here.
If we keep column names intact, then we need to surround them by quotes in general, so the first and last option seem to be identical to me.

Also, some grammar/style suggestions:
* `Keep Column Names intacts` should be `Keep column names intact` (no `s` on adjectives that apply to plural nouns unlike in French, plus standard capitalization of words in English)
* `Trim Column Names` -> "trimming" generally refers to removing whitespace at the beginning or the end. What we do is a more thorough normalization, so I would say `Normalize column names`
* `Use Escape Character for Column Names` -> what we have been discussing is the use of a *quote* character, not an escape one, although you will indeed need escaping if the quote character occurs in the string to be quoted

**Comment by the-star-sea** [comment]
Is there any other senior engineer to do the judge?It seems hard to do decisions efficiently here.

**Comment by the-star-sea** [comment]
Actually,the hardest thing is not how to do but what to do.I don't know what I should do and it is the second month now since I publish the pull request.

**Comment by antoine2711** [comment]
> I am struggling to understand why we need three options here. If we keep column names intact, then we need to surround them by quotes in general, so the first and last option seem to be identical to me.

Well @wetneb, it's to keep the possibility of having just the names, with no quotes or renaming. It's something we have now, so I didn't want to remove that possibility.

Thanks for the grammaticals corrections. I guess my french roots are still strong! ;-)

Regards, Antoine

**Comment by wetneb** [comment]
> Well @wetneb, it's to keep the possibility of having just the names, with no quotes or renaming. It's something we have now, so I didn't want to remove that possibility.

Hmm, unless I am mistaken I think what we have now is the second option (normalization introduced by a recent PR). Having just the names with no quotes or renaming can generate invalid SQL statements, so it is probably not an option we should offer to the user, should we?

**Comment by antoine2711** [comment]
> Actually,the hardest thing is not how to do but what to do.I don't know what I should do and it is the second month now since I publish the pull request.

@the-star-sea: I understand this is frustrating sometimes. The discussions with the community is a rather lengthy and uncertain process. But it's really what can make the tool better, as they use the product and as such, might come with perspective that a new developper migth not have think of.

I think Antonin recommandation to « Ideally, you should feel invited to take part to those design decisions, and have opinions about the features you are implementing, because you can imagine how users will interact with them. » is important, because doing so stimulate the discussions in the community, making the process faster.

Regards, Antoine

**Comment by antoine2711** [comment]
> Hmm, unless I am mistaken I think what we have now is the second option (normalization introduced by a recent PR). Having just the names with no quotes or renaming can generate invalid SQL statements, so it is probably not an option we should offer to the user, should we?

I think doing nothing (first option) is good for backward compatibility. I could also imagine a SQL engin that would allow UTF-8 field names, but not the double quote. I do agree that the 2 other options will probably satisfy 99% of the user base or popular SQL DB. So, definitely not a hill I would die for.

> * `Use Escape Character for Column Names` -> what we have been discussing is the use of a _quote_ character, not an escape one, although you will indeed need escaping if the quote character occurs in the string to be quoted

Well @wetneb, I think that I made this mistake from the start, even in the title of this issue… So, should it be called `Quote Column Names with (")`?

Regards, Antoine

**Comment by the-star-sea** [comment]
I think a good tool does not serve for all the people but for most of people.The more freedom,the less convenience.

**Comment by wetneb** [comment]
> I think doing nothing (first option) is good for backward compatibility. I could also imagine a SQL engin that would allow UTF-8 field names, but not the double quote. I do agree that the 2 other options will probably satisfy 99% of the user base or popular SQL DB. So, definitely not a hill I would die for.

Hmm, I really do not understand! Backwards compatibility is not about preserving existing bugs, right? The previous behaviour (before normalization of column names) was just incorrect. SQL engines have to support double quotes to quote column names, because that is specified by the standard. I cannot imagine a setting in which the user would enforce the generation of invalid SQL statements.

**Comment by antoine2711** [comment]
@the-star-sea: we now have a strong consensus about how to implement this features. Is it clear to you?

Regards, Antoine

**Comment by the-star-sea** [comment]
> Also, if we are indeed going for two options, then we can probably have just a checkbox rather than two radio buttons (no strong feelings about which way the checkbox should go: either normalizing when checked, or using quotes when checked - as long as it is consistent with the label of course…)

Yes, I am clear now.But these days I have 5 final exams coming soon to prepare.I will do it later.

**Comment by antoine2711** [comment]
> Yes, I am clear now.But these days I have 5 final exams coming soon to prepare.I will do it later.

No rush @the-star-sea. I just wanted to make sure you were confortable with the discussions and that you had what you needed to go forward. Good examens to you! 

Regards, Antoine

**Comment by Sudo-Aryan** [comment]
I have read and gone through each of discussion for this issue, I would like to give it a try

Regards, Aryan

**Comment by antoine2711** [comment]
> I have read and gone through each of discussion for this issue, I would like to give it a try

@Sudo-Aryan : you already have one « good first issues » assigned, we try to keep one per contributors. You can look at « [good second issues](https://github.com/OpenRefine/OpenRefine/labels/good%20second%20issue) ».

Regards, Antoine

**Comment by XxBrezxX** [comment]
Hi!, I wanna try this, is it still available?

**Comment by tfmorris** [comment]
I've removed the Good First Issue label until there's a better specification of what's needed here. I've also changed the terminology to use the conventional term "quote", rather than "escape", which is something different.

It looks like SQL has two different [classes of identifiers](https://learn.microsoft.com/en-us/sql/relational-databases/databases/database-identifiers?view=sql-server-ver16#classes-of-identifiers), regular identifiers and delimited identifiers, and there are (at least?) three different conventions for quoting "delimited identifiers":
```
1. "ANSI Standard"
2. [MS SQL Server]
3. `MySQL`
```

Our strategy for these has evolved over time:
1. Do nothing, the user is responsible for getting things correct
2. Attempt to transform column name into a legal regular identifier
3. Quote column name using the ANSI SQL quote character
4. ... whatever's next ...

If we're going to support non-ANSI standard quote characters, I'd suggest that we should support both SQL Server and MySQL (plus any other legacy proprietary quoting schemes which exist?), rather than just one of them.

**Comment by thadguidry** [comment]
As a power tool for non-programmers (and also non-DBAs typically), I think it makes sense to ease the burden and make our SQL export dialog even more clear when we do introduce the new quote schemes.  We can just use paratheses after the option to give a hint of what to use for which db vendor.  He's what the defaults look like for the vendors:

- [ ] Use ANSI quotes (PostgreSQL, SQL Server, Oracle, Snowflake, etc.)
- [ ] Use backtick quotes (MySQL, MariaDB, etc.)

NOTED:  Each db vendor has their own override pragmas (system) or options to change the acceptable quoting of identifiers.  But we should just target or terminology and handling of quote schemes to the db vendors defaults, which the community would expect from us I think.  "cater to the masses of non-programmers and make it easier for them to export their data to standard db setups"

**Comment by thadguidry** [comment]
It's worth noting that the change we made in 3.8.0 to normalize by always wrapping with double quotes `"` does break things a bit f[or users of MySQL in particular](https://forum.openrefine.org/t/trouble-with-sql-export/1485) which does not ship with [ANSI_QUOTES ](https://dev.mysql.com/doc/refman/8.4/en/string-literals.html)mode turned on and thus our enforcement of double quotes `"` around identifiers will cause an error in MySQL when executing the script, but no error if there's no quotes, or wrapped in back ticks.
